twitch.tv: Create a visibilityState Quirk to support play/pause of PIP videos when another tab is in the foreground

brentfulgham opened the pull request:
#### 54e267670df1b6c7e018bcf853d798fad33c2577
<pre>
twitch.tv: Create a visibilityState Quirk to support play/pause of PIP videos when another tab is in the foreground
<a href="https://bugs.webkit.org/show_bug.cgi?id=303674">https://bugs.webkit.org/show_bug.cgi?id=303674</a>
&lt;<a href="https://rdar.apple.com/problem/102420527">rdar://problem/102420527</a>&gt;

Reviewed by Jer Noble and Youenn Fablet.

Some media players use the VisibilityState API as part of their media controls logic. This
creates a problem for some PIP flows, where the PIP video might be owned by a tab the user
has moved away from.

Since switching from one tab to another while in PIP is a very common use case, this patch
creates a quirk that adjusts the behavior of visibilityState to indicate that the document
is visible if it has an active PIP window.

Test: Tools/TestWebKitAPI/Tests/WebKitCocoa/VisibilityState.mm

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::visibilityState const): Check if the current document requires the
new ShouldReportDocumentAsVisibleIfActivePIPQuirk. If it does, and the document has an
active PIP window, report the document as visible.
* Source/WebCore/page/Quirks.cpp:
(WebCore::handleTwitchQuirks): Add quirk for twitch.tv
(WebCore::Quirks::determineRelevantQuirks): Add quirk for twitch.tv.
* Source/WebCore/page/Quirks.h:
(WebCore::Quirks::shouldReportVisibleDueToActivePictureInPictureContent const): Added.
* Source/WebCore/page/QuirksData.h: Add new ShouldReportDocumentAsVisibleIfActivePIPQuirk
* Tools/TestWebKitAPI/Tests/WebKitCocoa/VisibilityState.mm:
(TEST(VisibilityState, PIPKeepsDocumentVisibleQuirk)): Added.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/twitch_quirk.html: Added.

Canonical link: <a href="https://commits.webkit.org/304308@main">https://commits.webkit.org/304308@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a54d851dc812ea46ece88c32d8bfad760c34cd79

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/135209 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/7601 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/46475 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/142716 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/86976 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 ios-apple 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/137078 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/8237 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/7449 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/103323 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/86976 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/138155 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/5876 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/121190 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/84182 "Passed tests") | | ⏳ 🛠 vision-apple 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/5666 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/3269 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/3306 "Built successfully") | | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/114869 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/39365 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/145412 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/7283 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/39933 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/111703 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/7327 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/6101 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/112067 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28432 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/5510 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/117484 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/61234 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/7337 "Built successfully") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/35618 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/7093 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/70889 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/7313 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/7196 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

jernoble commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -2707,6 +2707,12 @@ VisibilityState Document::visibilityState() const
         return VisibilityState::Hidden;
+#if ENABLE(PICTURE_IN_PICTURE_API)
+    if (quirks().shouldReportVisibleDueToActivePictureInPictureContent()) {
+        if (pictureInPictureElement())
```
This can be collapsed to a single `if()`:
```suggestion
    if (quirks().shouldReportVisibleDueToActivePictureInPictureContent() && pictureInPictureElement())
```

brentfulgham replied:
Will do!

jernoble commented on Source/WebCore/loader/DocumentWriter.cpp:
```diff
@@ -250,6 +251,13 @@ bool DocumentWriter::begin(const URL& urlReference, bool dispatch, Document* own
+        if (!topURLForTesting.isNull())
+            document->quirks().setTopDocumentURLForTesting(WTFMove(topURLForTesting));
+    }
+
```
Is this needed for this patch?

brentfulgham replied:
If we set a Test URL before the page has parsed (i.e., start a load request, then tell the web view to use a specific domain for testing quirks) it gets lost -- so this resolves that.

But I think I can avoid changing this code by revising the test case to make this in the web content.

jernoble commented on Source/WebCore/page/Quirks.h:
```diff
@@ -196,6 +196,7 @@ class Quirks {
 #endif
 
     WEBCORE_EXPORT void setTopDocumentURLForTesting(URL&&);
+    URL copyTopDocumentURLForTesting() const { return m_topDocumentURLForTesting; }
```
...ditto.

jernoble reviewed (approved):
r=me with nits

youennf commented on Source/WebCore/page/Quirks.cpp:
```diff
@@ -2407,6 +2407,17 @@ bool Quirks::shouldExposeCredentialsContainerQuirk() const
+}
+#endif
+
+
```
```suggestion
```

brentfulgham committed:
twitch.tv: Create a visibilityState Quirk to support play/pause of PIP videos when another tab is in the foreground
https://bugs.webkit.org/show_bug.cgi?id=303674
<rdar://problem/102420527>

Reviewed by Jer Noble and Youenn Fablet.

Some media players use the VisibilityState API as part of their media controls logic. This
creates a problem for some PIP flows, where the PIP video might be owned by a tab the user
has moved away from.

Since switching from one tab to another while in PIP is a very common use case, this patch
creates a quirk that adjusts the behavior of visibilityState to indicate that the document
is visible if it has an active PIP window.

Test: Tools/TestWebKitAPI/Tests/WebKitCocoa/VisibilityState.mm

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::visibilityState const): Check if the current document requires the
new ShouldReportDocumentAsVisibleIfActivePIPQuirk. If it does, and the document has an
active PIP window, report the document as visible.
* Source/WebCore/page/Quirks.cpp:
(WebCore::handleTwitchQuirks): Add quirk for twitch.tv
(WebCore::Quirks::determineRelevantQuirks): Add quirk for twitch.tv.
* Source/WebCore/page/Quirks.h:
(WebCore::Quirks::shouldReportVisibleDueToActivePictureInPictureContent const): Added.
* Source/WebCore/page/QuirksData.h: Add new ShouldReportDocumentAsVisibleIfActivePIPQuirk
* Tools/TestWebKitAPI/Tests/WebKitCocoa/VisibilityState.mm:
(TEST(VisibilityState, PIPKeepsDocumentVisibleQuirk)): Added.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/twitch_quirk.html: Added.

Canonical link: https://commits.webkit.org/304308@main
